bad partition key when re-keying table

### Hail version:
devel-10a75bb57a6f
### What you did:

```
def test_rekey_correct_partition_key(self):
        ht = hl.utils.range_table(5)
        ht = ht.add_index('a')
        ht = ht.key_by('idx', 'a')
        ht = ht.annotate(b=ht.idx)
        ht = ht.key_by('idx', 'b')
        self.assertEqual(ht.aggregate(agg.sum(ht.idx)), 10)
```

### What went wrong (all error messages here, including the full java stack trace):

```
Java stack trace:
java.lang.AssertionError: assertion failed
	at scala.Predef$.assert(Predef.scala:156)
	at is.hail.rvd.OrderedRVDType.<init>(OrderedRVDType.scala:21)
	at is.hail.rvd.OrderedRVDType.copy(OrderedRVDType.scala:121)
	at is.hail.expr.TableKeyBy.execute(Relational.scala:1857)
	at is.hail.expr.TableMapRows.execute(Relational.scala:2090)
	at is.hail.expr.TableKeyBy.execute(Relational.scala:1846)
	at is.hail.expr.ir.Interpret$.apply(Interpret.scala:520)
	at is.hail.expr.ir.Interpret$.apply(Interpret.scala:39)
	at is.hail.table.Table.aggregate(Table.scala:465)
	at is.hail.table.Table.aggregate(Table.scala:446)
	at is.hail.table.Table.aggregateJSON(Table.scala:436)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at py4j.reflection.MethodInvoker.invoke(MethodInvoker.java:244)
	at py4j.reflection.ReflectionEngine.invoke(ReflectionEngine.java:357)
	at py4j.Gateway.invoke(Gateway.java:280)
	at py4j.commands.AbstractCommand.invokeMethod(AbstractCommand.java:132)
	at py4j.commands.CallCommand.execute(CallCommand.java:79)
	at py4j.GatewayConnection.run(GatewayConnection.java:214)
	at java.lang.Thread.run(Thread.java:748)
```

--- a/python/hail/tests/test_api.py
+++ b/python/hail/tests/test_api.py
@@ -620,12 +620,17 @@ class TableTests(unittest.TestCase):
         self.assertEqual(ht.aggregate(hl.agg.array_sum(ht.x) / [2, 2]), [5.0, 10.0])
 
     def test_rekey_correct_partition_key(self):
-            ht = hl.utils.range_table(5)
-            ht = ht.add_index('a')
-            ht = ht.key_by('idx', 'a')
-            ht = ht.annotate(b=ht.idx)
-            ht = ht.key_by('idx', 'b')
-            self.assertEqual(ht.aggregate(agg.sum(ht.idx)), 10)
+        ht = hl.utils.range_table(5)
+        ht = ht.annotate(a = ht.idx)
+        ht = ht.key_by('idx', 'a')
+        ht = ht.key_by('idx')
+        ht._force_count()
+        
+        ht = hl.utils.range_table(5)
+        ht = ht.annotate(a = ht.idx, b = ht.idx)
+        ht = ht.key_by('idx', 'a')
+        ht = ht.key_by('idx', 'b')
+        ht._force_count()
 
 
 class MatrixTests(unittest.TestCase):

--- a/python/hail/tests/test_api.py
+++ b/python/hail/tests/test_api.py
@@ -619,6 +619,15 @@ class TableTests(unittest.TestCase):
         ht = ht.annotate(x = [1,2])
         self.assertEqual(ht.aggregate(hl.agg.array_sum(ht.x) / [2, 2]), [5.0, 10.0])
 
+    def test_rekey_correct_partition_key(self):
+            ht = hl.utils.range_table(5)
+            ht = ht.add_index('a')
+            ht = ht.key_by('idx', 'a')
+            ht = ht.annotate(b=ht.idx)
+            ht = ht.key_by('idx', 'b')
+            self.assertEqual(ht.aggregate(agg.sum(ht.idx)), 10)
+
+
 class MatrixTests(unittest.TestCase):
     def get_vds(self, min_partitions=None) -> hl.MatrixTable:
         return hl.import_vcf(resource("sample.vcf"), min_partitions=min_partitions)

--- a/src/main/scala/is/hail/expr/Relational.scala
+++ b/src/main/scala/is/hail/expr/Relational.scala
@@ -1852,8 +1852,8 @@ case class TableKeyBy(child: TableIR, keys: Array[String], nPartitionKeys: Optio
 
       tv.rvd match {
         case ordered: OrderedRVD =>
-          if (ordered.typ.key.startsWith(keys)
-            && nPartitionKeys.forall(n => ordered.typ.partitionKey.length == n))
+          if (ordered.typ.key.startsWith(keys) &&
+            nPartitionKeys.getOrElse(keys.length) == ordered.typ.partitionKey.length)
             ordered.copy(typ = ordered.typ.copy(key = keys))
           else resort
         case _: UnpartitionedRVD =>

